MSVC and  -DCMAKE_C_STANDARD=23

zlib-ng fails to compile with MSVC (latest version) and  -DCMAKE_C_STANDARD=23
compile error:
```
deflate.c(714,9): warning C4013: 'unreachable' undefined; assuming extern returning int
```
The reason is that unreachable is not defined in stddef.h (latest Windows SDK, as well as latest Windows Insider SDK) yet.

A possible workaround for zbuild.h:
```cpp
/* Hint to compiler that a block of code is unreachable, typically in a switch default condition */
#ifndef Z_UNREACHABLE
#  if defined(__STDC_VERSION__) && __STDC_VERSION__ >= 202311L
#    if defined(_MSC_VER)
#      define Z_UNREACHABLE() __assume(0)
#    else
#      define Z_UNREACHABLE() unreachable()           // C23 approach
#    endif
#  elif (defined(__GNUC__) && (__GNUC__ >= 5)) || defined(__clang__)
#    define Z_UNREACHABLE() __builtin_unreachable()
#  else
#    define Z_UNREACHABLE()
#  endif
#endif
```

--- a/zbuild.h
+++ b/zbuild.h
@@ -53,7 +53,11 @@
 /* Hint to compiler that a block of code is unreachable, typically in a switch default condition */
 #ifndef Z_UNREACHABLE
 #  if defined(__STDC_VERSION__) && __STDC_VERSION__ >= 202311L
-#    define Z_UNREACHABLE() unreachable()           // C23 approach
+#    if !defined(unreachable) && defined(_MSC_VER)
+#      define Z_UNREACHABLE() __assume(0)
+#    else
+#      define Z_UNREACHABLE() unreachable()           // C23 approach
+#    endif
 #  elif (defined(__GNUC__) && (__GNUC__ >= 5)) || defined(__clang__)
 #    define Z_UNREACHABLE() __builtin_unreachable()
 #  else

